### ISSUE TITLE
Consider not throwing when refresh token is not set

### ISSUE THREAD

**Issue by richardson-trevor** [google/google-auth-library-nodejs#290]
#### Context
We are currently building an http-triggered cloud func with auth based on the Google Solutions example show [here](https://cloud.google.com/solutions/authentication-in-http-cloud-functions). In our case, we have a service running in app engine (using default credential chain) calling a cloud func with its token in the authorization request header. In order to make sure the caller is authenticated, we use the Storage.testIamPermissions api. 

#### Problem
This all works properly from our app that is authenticated, but when something calls us that is not authenticated we get an error with: `No refresh token set` from [here in the auth library](https://github.com/google/google-auth-library-nodejs/blob/78d6813ef648b32312b01396e782f234599ed706/src/auth/oauth2client.ts#L481). 

#### Question / Request for guidance
Is there a better way to handle this outside a try/catch with a specific catch that looks for this error? Something I'd be looking for would be a valid response with details declaring that the request was not authenticated and potentially a "reason" saying `No refresh token set` rather than a thrown error.

**Comment by JustinBeckwith** [comment]
Greetings!  Could you share the auth code inside your cloud function?  It would be super helpful to see exactly how you're using the API.

**Comment by richardson-trevor** [comment]
Appreciate the timely response! Sure, here's what we're doing:
```js
const oauth = new google.auth.OAuth2();
oauth.setCredentials({access_token: accessToken});

const bucket = google.storage('v1').buckets;
const testIamPermissions = promisify(bucket.testIamPermissions);
const response = await testIamPermissions({bucket: BUCKET, permissions: [permission], auth: oauth}, {});
```

**Comment by JustinBeckwith** [comment]
Alright, I know what's going on here.  This is (yet another) bug that's brought about from retrying failed requests.  This is the culprit:
https://github.com/google/google-auth-library-nodejs/blob/master/src/auth/oauth2client.ts#L703

When you pass an invalid access_token (which is essentially what your code is doing), the request should fail with an error like this:

```
{ Error: Invalid Credentials
    at createError (/Users/beckwith/Code/google-auth-library-nodejs/node_modules/axios/lib/core/createError.js:16:15) ....
```

However, right now we swallow all errors that come back from the `request<T>` method, and automatically try refreshing the token.  In this case...  we try to refresh the token even if no refresh token has been configured 😐  

```
Error: No refresh token is set.
    at OAuth2Client.<anonymous> (/Users/beckwith/Code/google-auth-library-nodejs/build/src/auth/oauth2client.js:242:35)
```

The answer here is to just eliminate the retry inside of the `request<T>` method, as it does nothing but cause trouble. We've discussed doing this in the past, but we're worried that it's a semver breaking change.  For now I'm going to attach a semver major label to this one while we discuss the right way to fix it with minimal impact.

**Comment by JustinBeckwith** [comment]
Also, here's a minimal repro from our existing samples:

```js
const { OAuth2Client } = require('google-auth-library');
async function main() {
  const oAuth2Client = new OAuth2Client();
  oAuth2Client.setCredentials({ access_token: 'Bearer notarealtoken'});
  const url = 'https://www.googleapis.com/plus/v1/people?query=pizza';
  const res = await oAuth2Client.request({ url });
  console.log(res.data);
}
main().catch(console.error);
```

**Comment by JustinBeckwith** [comment]
@ofrobots FYI

**Comment by richardson-trevor** [comment]
Agreed; sounds spot-on as to what's going on. We are indeed guarding against the empty possibility. We ended up running into it because we misunderstood an iam role's permission on app engine. 

Greatly appreciate the quick feedback. Will keep my eye out to see what you guys move forward with. 

If you'd be open to the consideration, my vote would actually be to have the client only fail with an error when the request itself blows up somehow. In cases where you simply fail to authenticate, but the verification process goes through properly, I'd prefer the api actually respond with a valid response, but contains an indicator of `authenticated: false` or similar. In my mind it'd be similar to a request api returning a successful promise on a 400-500 range response since a valid http response did indeed come back properly.

**Comment by JustinBeckwith** [comment]
With all of these calls, failed authentication is the same as the request failing.  We don't do a separate `authenticate` and `request` call.  We're just taking the access_token you give us, base64 encoding it, and adding it as an authorization header.  That's pretty much it :)  The API you're calling checks the header, determines the access is invalid, and returns a [401](https://httpstatuses.com/401).  Nothing about that is likely to change. 

Now the point of discussion could be the desire to throw on a non-200 response status code vs returning a valid response object where you have to check the status code.  This is really a matter of taste, and one of the things [axios](https://github.com/axios/axios) and [request](https://github.com/request/request) does differently.  Axios, which we now use in this library, assumes any non-200 response is an error (which is generally true), while Request does not return an error in the callback unless there was a network issue (ETIMEDOUT, ENOTFOUND, etc).

Given the fact that we're using Axios, I'd like to keep the current pattern.  If you prefer a pattern more in line with request or fetch, you can always just use the `getRequestMetadata` method to obtain the authorization headers, and take control of the final request flow.  Arguably, that's a better way to use this library anyhow :)

**Comment by richardson-trevor** [comment]
Thanks for taking the time to explain the details here. I will take a look at `getRequestMetadata`. 

Thanks again, and keep up the good work! This stuff is really helping our team.

**Comment by JustinBeckwith** [comment]
Resolved by #307
